contracts/JumpBall.cdc

```diff
@@ -0,0 +1,379 @@
+            game.transferAllToWinner(winner: winner, winnerCap: winnerCap)
+        }
+    }
+
+    // Factory function to create a new Player resource
+    access(all) fun createPlayer(playerAddress: Address, collectionCap: Capability<&{NonFungibleToken.Collection}>): @Player {
+        if !collectionCap.check() {
+            panic("A valid NFT collection capability is required to create a player.")
+        }
+
+        return <- create Player(
+            address: playerAddress,
+            collectionCap: collectionCap
+        )
+    }
+
+    access(all) var metadata: {String: AnyStruct}
+    access(all) var games: @{String: Game}
+    access(all) var userGames: {Address: [String]}
+    access(all) let admin: @Admin
+
+    init() {
+        self.metadata = {}
+        self.admin <- create Admin()
+        self.games <- {}
+        self.userGames = {}
+
+        // Initialize Player resource for contract owner
+        let collectionCap = self.account.capabilities.storage.issue<&{NonFungibleToken.Collection}>(/storage/OwnerNFTCollection)
+        self.account.capabilities.publish(collectionCap, at: /public/OwnerNFTCollection)
+
+        // Save the Player resource in storage
+        self.account.storage.save<@Player>(<- create Player(
+            address: self.account.address,
+            collectionCap: collectionCap
+        ), to: /storage/JumpBallPlayer)
+    }
+
+    // Helper functions for contract metadata
+    access(all) fun setMetadata(key: String, value: AnyStruct) {
+        self.metadata[key] = value
+    }
+
+    access(all) fun getMetadata(key: String): AnyStruct? {
+        return self.metadata[key]
+    }
+
+    // Add a game to a user's list of games
+    access(self) fun addGameForUser(user: Address, gameID: String) {
+        if JumpBall.userGames[user] == nil {
+            JumpBall.userGames[user] = []
+        }
+        JumpBall.userGames[user]?.append(gameID)
+    }
+
+    // Retrieve all games for a given user
+    access(all) fun getGamesByUser(user: Address): [String] {
+        return JumpBall.userGames[user] ?? []
+    }
+
+    // Get a reference to a specific game
+    access(all) fun getGame(gameID: String): &Game? {
+        return &JumpBall.games[gameID] as &Game?
+    }
+
+    access(all) fun gameExists(gameID: String): Bool {
+        return JumpBall.games[gameID] != nil
+    }
+
+    // Handle timeout: allows users to reclaim their moments
+    access(all) fun claimTimeout(gameID: String, claimant: Address) {
+        let gameRef = &JumpBall.games[gameID] as &Game?
+            ?? panic("Game does not exist.")
+
+        if JumpBall.getCurrentTime() < gameRef.startTime + gameRef.gameDuration {
+            panic("Game is still in progress.")
+        }
+
+        emit TimeoutClaimed(gameID: gameID, claimant: claimant)
+        gameRef.returnAllNFTs()
+    }
+
+    // Destroy a game and clean up resources
+    access(all) fun destroyGame(gameID: String) {
```

Comment:
This can also be called publicly, maybe this should be part of the Player resource with logic to validate the creator/opponent address? And/or this could be part of the Admin resource?